lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+        };
+
+        // e.decompress consumes e, so we need to call it separately, after populating the fields
+        // above.
+        let data = if e.is_compress {
+            e.decompress()?
+        } else {
+            e.data
+        };
+        res.data_ptr = data.as_ptr();
+        res.data_len = data.len();
+
+        Ok(res)
+    }
+}
+
+// TODO: this is Linux-specific, also implement for Windows.
+fn wait_for_fd(fd: usize) -> bool {
+    unsafe {
+        let mut raw_fds: fd_set = mem::zeroed();
+
+        FD_SET(fd as i32, &mut raw_fds);
+
+        let result = select(
+            fd as i32 + 1,
+            &mut raw_fds,
+            ptr::null_mut(),
+            ptr::null_mut(),
+            ptr::null_mut(),
+        );
+        result == 1
+    }
+}
+
+// read_rdp_output reads incoming RDP bitmap frames from client at client_ref and forwards them to
+// handle_bitmap. handle_bitmap *must not* free the memory of CGOBitmap.
+#[no_mangle]
+pub extern "C" fn read_rdp_output(
+    client_ref: i64,
+    handle_bitmap: unsafe extern "C" fn(i64, CGOBitmap) -> CGOError,
+) -> CGOError {
+    let mut tcp_fd = 0;
+    if let Err(e) = with_client(&client_ref, |client| -> Result<(), ClientError> {
+        tcp_fd = client.lock().unwrap().tcp_fd;
+        Ok(())
+    }) {
+        return format!(
+            "failed looking up TCP file descriptor for client {}: {:?}",
+            client_ref, e
+        )
+        .into();
+    }
+    // Read incoming events.
+    // TODO: this doesn't always unblock after client.shutdown() was called. Figure out why.
+    while wait_for_fd(tcp_fd as usize) {
+        if let Err(e) = with_client(&client_ref, |client| -> Result<(), ClientError> {
+            let mut err = CGOError::null();
+            let mut client = client.lock().unwrap();
+            client.rdp_client.read(|rdp_event| match rdp_event {
+                RdpEvent::Bitmap(bitmap) => {
+                    let cbitmap = match CGOBitmap::try_from(bitmap) {
+                        Ok(cb) => cb,
+                        Err(e) => {
+                            println!(
+                                "failed to convert RDP bitmap to CGO representation: {:?}",
+                                e
+                            );
+                            return;
+                        }
+                    };
+                    unsafe {
+                        err = handle_bitmap(client_ref, cbitmap);
+                    };
+                }
+                // These should never really be sent by the server to us.
+                RdpEvent::Pointer(_) => {
+                    println!("got unexpected pointer event from RDP server, ignoring");
+                }
+                RdpEvent::Key(_) => {
+                    println!("got unexpected keyboard event from RDP server, ignoring");
+                }
+            })?;
+            if !err.is_null() {
+                Err(ClientError::Other(format!(
```

Comment:
Ideally you should define explicit enum variants with conversions and whatnot for these errors and not use weakly typed strings.

Reply:
Maybe https://docs.rs/err-derive/0.3.0/err_derive/?

Reply:
`thiserror` is generally the recommend derive crate for this. The downside is that all derive crates incur lots of compile time bloat and smaller amounts of runtime size bloat. For small applications like this I generally just spend the required 5 minutes writing them.